src/CoreLib/DSetCSharpLinq.fs

```diff
@@ -0,0 +1,104 @@
+    [<Extension>]
+    static member Where(dset : DSet<'U>, func : Func<'U, bool>) =
+        dset.Filter(func)
+
+    /// Fold the entire DSet with a fold function, an aggregation function, and an initial state. The initial state is deserialized (separately) for each partition. 
+    /// Within each partition, the elements are folded into the state variable using 'folder' function. Then 'aggrFunc' is used to aggregate the resulting
+    /// state variables from all partitions to a single state.
+    [<Extension>]
+    static member Aggregate(dset : DSet<'U>, state : 'State, folder : Func<'State, 'U, 'State>, aggrFunc : Func<'State, 'State, 'State>) =
+        dset.Fold(folder, aggrFunc, state)
+
+     /// Reduces the elements using the specified 'reducer' function
+    [<Extension>]
+    static member Aggregate(dset : DSet<'U>, reducer : Func<'U, 'U, 'U>) =
+        dset.Reduce(reducer)
+
+    /// Groups the elements of a dataset according to a specified key selector function.
+    [<Extension>]
+    static member GroupBy(dset : DSet<'U>, keySelector : Func<'U, 'K>) =
+        dset.Map(fun u -> (keySelector.Invoke(u), u)).GroupByKey()
+        
+    /// Groups the elements of a dataset according to a specified key selector function and projects the elements for each group by using a specified function.
+    [<Extension>]
+    static member GroupBy(dset : DSet<'U>, keySelector : Func<'U, 'K>, elementSelector : Func<'U, 'U1>) =
+        DSetLinqExtensions.GroupBy(dset, keySelector).Map(fun (_, grp) -> grp |> Seq.map(fun v -> elementSelector.Invoke(v)))
+        
+    /// Groups the elements of a dataset according to a specified key selector function and creates a result value from each group and its key.
+    [<Extension>]
+    static member GroupBy(dset : DSet<'U>, keySelector : Func<'U, 'K>, resultSelector : Func<'K, IEnumerable<'U>, 'U1>) =
+        DSetLinqExtensions.GroupBy(dset, keySelector).Map(fun (k, grp) -> resultSelector.Invoke(k, grp))
+
+    /// Groups the elements of a dataset according to a specified key selector function and creates a result value from each group and its key. The elements of each group are projected by using a specified function.
+    [<Extension>]
+    static member GroupBy(dset : DSet<'U>, keySelector : Func<'U, 'K>, elementSelector : Func<'U, 'U1>, resultSelector : Func<'K, IEnumerable<'U1>, 'U2>) =
+        DSetLinqExtensions.GroupBy(dset, keySelector).Map(fun (k, grp) -> k, grp |> Seq.map(fun v -> elementSelector.Invoke(v))).Map(fun (k, grp) -> resultSelector.Invoke(k, grp) )
+
+    /// Returns the dataset as an IEnumerable<'U>
+    [<Extension>]
+    static member AsEnumerable(dset : DSet<'U>) = 
+        dset.ToIEnumerable()
+
+    /// Produces the set union of two datasets
+    [<Extension>]
+    static member Union(dset1 : DSet<'U>, dset2 : DSet<'U> ) =
+        dset1.Union(seq { yield dset2 })
```

Comment:
Note: Linq's Union uses default equality comparer for type 'U.

For Prajna's, as it is now, Union merges all records without comparison to remove duplicates. We may want to revisit the semantics of Union.